server/src/main/java/org/apache/druid/discovery/DruidNodeDiscoveryProvider.java

```diff
@@ -139,45 +150,39 @@ public void nodesAdded(List<DiscoveryDruidNode> nodesDiscovered)
             }
           }
 
-          ImmutableList<DiscoveryDruidNode> newNodesAdded = null;
-          if (uninitializedNodeTypeListeners.isEmpty()) {
-            newNodesAdded = builder.build();
-          } else if (uninitializedNodeTypeListeners.remove(this) && uninitializedNodeTypeListeners.isEmpty()) {
```

Comment:
Right after it's initialized we propagate all accumulated `nodes` to the listener (old version), in the new version we don't. Could we make comments to the class which explains the purpose of the new behaviour, why it's better?

Reply:
I don't understand your question. Could you suggest what Javadoc comment you think should be added and where?

Note: it shouldn't explain anything about the "old version", because that code won't exist anymore.

Reply:
It's unclear why we removed `uninitializedNodeTypeListeners.remove(this) && uninitializedNodeTypeListeners.isEmpty()`, was it redundant? If it was, it's unclear why.

Reply:
It was not redundant, it was correct, but extremely convoluted logic.

Reply:
Why I removed it: because it was extremely convoluted. I tried to replace that logic with something more comprehensible (hopefully).

Reply:
So, (in old version) it tried to propagate all accumulated nodes right after `initialized`, and (now) it propagates all (non-empty) events, right?